README.md


## go-loadgen
![logo](logo.png)

go-loadgen is a log load generator which is meant for testing the logging infrastructure. It is capable of producing structured or unstructured/structure logs with random alphanumeric strings and can also playback any input file at a specified controlled rate. 
## Build
This program requires GO 1.7. 

`go build -o loadgen main.go`

If you prefer to use vendor, then you can use the below command
`go build -v -o loadgen main.go`

## Usage
Use the help command to find all available commands and flags. 

    HHNLWB76D9289E:go-loadgen cbhatt1$ ./loadgen --help
    loadgen is a tool which generates test-data (synthetic/replay) at a controlled rate to test logging infrastructure.
    
    Usage:
      loadgen [COMMANDS] [flags]
      loadgen [command]
    
    Available Commands:
      help           Help about any command
      random-strings test data is generated by producing random alphanumeric strings.
      replay-file    test data is generated by replaying a file.
    
    Flags:
          --config string                          -config /path/to/source/jsonfile
          --duration int                           Duration of the job in seconds (default 300)
          --enable-log-rotation                    by default log rotation is off, use this flag to turn it on. Example --enable-log-rotation will turn log rotation on
          --file-count int                         target requests/sec. Note: log rotation does not work in multi file scenario. (default 1)
      -h, --help                                   help for loadgen
          --lines-per-second int                   target lines/second (default 1000)
          --output string                          file path of output log file ex: /var/tmp/test.log (default "default.log")
          --log-rotation-max-size-megabytes int   Rotate after N MegaBytes. example 100.  (default 100)
    
    Use "loadgen [command] --help" for more information about a command.

**Example-1:** To start a test with random-strings in the log file
Most of the local flags to this command is mandatory.

> The command generates 5 lines/second, with each line of length
> 1KiloBytes (1000 ASCII characters) and log entry spanning not more
> than 1 line (line-count) per event (no line breaks).
> The command produces a log file which grows at a rate of 5 KiloBytes/second and runs for 5 seconds. The overall size of the file after the test completes will be equal to 5 KB*5 seconds = 25KB

    ./loadgen random-strings --file-count 1 \
    --duration 5 \
    --enable-log-rotation \
    --line-count 1 \
    --lines-per-second 5 \
    --multi-line-percent 0 \
    --line-length  1000 \
    --output /var/log/test-alphanumeric.log


**Example-2:** To start a test with random-strings in the log file with 100% of the log entries containing multi-line.

> The flag --line-count and --multi-line-percent determines the number of line-breaks in the multi-line log entry. The resulting log file will be written at a byte rate of ***line-length\*line-count\*lines-per-second*** = 1000\*5\*5 = 25000 bytes/second. 100% of the log entries in the resulting log file contains line-break character at the end of each line, with each line spanning 1000 characters in length

    ./loadgen random-strings --file-count 1 \
    --duration 5 \
    --enable-log-rotation \
    --line-count 5 \
    --lines-per-second 5 \
    --multi-line-percent 100 \
    --line-length  1000 \
    --output /var/log/test-alphanumeric.log

## File Replay/File Playback
To start a log replay test with a specified input file.
> Note: Local flags such as duration and lines-per-second is required.

    ./loadgen replay-file \
    --input-file-path /Users/cbhatt1/git-logging/go-loadgen/test-data/sbg-pp-data.txt \
    --output test.log\
    --lines-per-second 1 \
    --duration 5

You can also specify the number of times a file should be replayed using the `--replay-count` flag.  

> If you use the `--replay-count` flag, then the flag `--duration` cannot be used, since the
> former flag becomes the test termiantion criteria and you cannot simultenously use 2 different termination criteria for the test.

In the below example, the test will complete after replaying the file "5" times.

Example:

    ./loadgen replay-file \
    --input-file-path /Users/cbhatt1/git-logging/go-loadgen/test-data/sbg-pp-data.txt \
    --output test.log\
    --lines-per-second 1 \
    --replay-count 5


## Customized fields using tags
go-loadgen can introduce custom fields using the tags flag. Provide tags in the form of `K=V` and those tags will be included in the log event. This particularly useful when metadata such as testId and/or line count should be present in the logs for backend validation.

     ./loadgen replay-file \
    --input-file-path /Users/cbhatt1/git-logging/go-loadgen/test-data/sbg-pp-data.txt \
    --output test.log\
    --enable-log-rotation \
    --log-rotation-max-size-megabytes 10 \
    --lines-per-second 1 \
    --file-count 3 \
    --duration 5 \
    --tags "lineCount=1 testId=1"

## Defining a test results log file
go-loadgen can generate a test results log file in the specified path. `--result-log` is the persistent flag which enables that. Use it as shown below. 
It generates a file with "total_lines_generated=x" entry which can be used as a way to assert whether the generated load is equal to the load received on the backend.

     ./loadgen replay-file \
    --input-file-path /Users/cbhatt1/git-logging/go-loadgen/test-data/sbg-pp-data.txt \
    --output test.log\
    --enable-log-rotation \
    --log-rotation-max-size-megabytes 10 \
    --lines-per-second 1 \
    --file-count 1 \
    --duration 5 \
    --result-log /var/tmp/result.log"
    

## Log rotation and file count
For certain tests, it is critical to generate multiple log files along with log-rotation and backup file retention policies. Here is how you can achieve that using go-loadgen:

     ./loadgen replay-file \
    --input-file-path /Users/cbhatt1/git-logging/go-loadgen/test-data/sbg-pp-data.txt \
    --output test.log\
    --enable-log-rotation \
    --log-rotation-max-size-megabytes 10 \
    --lines-per-second 1 \
    --file-count 3 \
    --duration 5
    
In the above CLI command : 

 - The argument `--enable-log-rotation` enables log-rotation
 -   The argument pair `--rotation-max-file-size-megabytes 10` specifies how
   large a file can grow before the file is rotated with a new one.  In
   this example the limit is 10MB.
 - The argument pair `--file-count 3` determines how many log files the test has to create and update simultenously. 

go-loadgen currently allows for a maximum of 3 backup files after which the oldest file will be deleted. 

Log rotation default settings

|  Property| Default | Comments |
|--|:--:|:--|
| MaxFileSize |  500MB| Maximum size the file is allowed to grow|
| MaxAge |  1 Day|After MaxAge exceeds the file is rotated automatically even if size is within the max limit.|
| MaxBackups |  3|Number of rotated files.|
| File Compression |  Disabled| |



## Prometheus Metrics

To start a test with prometheus metrics exposed to ```http://localhost:<port>/metrics``` endpoint, use the "enable-metrics" global flag
Note: the default prometheus listener port is "8080" unless specified using the *`--metrics-server-port <port>`* flag
   

     ./loadgen replay-file \
        --input-file-path /Users/cbhatt1/git-logging/go-loadgen/test-data/sbg-pp-data.txt \
        --output test.log\
        --lines-per-second 1 \
        --duration 5 \
        --enable-metrics \
        --metrics-server-port 8081

Example output using curl:

     cbhatt1$ curl http://localhost:8081/metrics
    # HELP total_bytes_processed Total bytes processed
    # TYPE total_bytes_processed counter
    total_bytes_processed 2.23982e+07
    # HELP total_events_processed Total log events generated by the tool.
    # TYPE total_events_processed counter
    total_events_processed 13177


## License ##

See `LICENSE` for details


CHANGELOG.md

<a name="v1.0.2"></a>
## [v1.0.2](y/compare/v1.0.1...v1.0.2) (2020-03-03)


<a name="v1.0.1"></a>
## [v1.0.1](y/compare/v1.0.0...v1.0.1) (2020-03-03)


<a name="v1.0.0"></a>
## v1.0.0 (2020-03-03)



cli/cli.go
package cli

import (
	"errors"
	"fmt"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/spf13/cobra"
	easy "github.com/t-tomalak/logrus-easy-formatter"
	"github.com/intuit/go-loadgen/constants"
	loadgen "github.com/intuit/go-loadgen/loadgenerator"
	"github.com/intuit/go-loadgen/metrics"
	"os"
	"sync"
)

const (
	lineCount                   = "line-count"
	lineLength                  = "line-length"
	fileCount                   = "file-count"
	duration                    = "duration"
	enableRotate                = "enable-log-rotation"
	linesPerSecond              = "lines-per-second"
	multiLinePercent            = "multi-line-percent"
	logRotationMaxSizeMegabytes = "log-rotation-max-size-megabytes"
	enableMetrics               = "enable-metrics"
	metricsServerPort           = "metrics-server-port"
	inputFilePath               = "input-file-path"
	replayCount                 = "replay-count"
	output                      = "output"
	config                      = "config"
	tags                        = "tags"
	enableConfig                = "enable-config"
	resultLog                   = "result-log"
)

func InputValidation(props *loadgen.LoadGenProperties) {

	if ! props.Rotate {
		fmt.Println("Warning. File rotation is disabled. You may run out of disk space if the file grows too large!")
	}
	if props.FileCount > constants.MaxFileCount {
		fmt.Printf("Too many files in --file-count parameter. This program can only support upto %d files\n", constants.MaxFileCount)
		os.Exit(1)
	}
}

var cfgFile string

func printInput(props *loadgen.LoadGenProperties) {
	fmt.Printf("Input values = ")
	fmt.Printf(lineCount+" = %d\n", props.NumOfLinesInMultiLineLog)
	fmt.Printf(lineLength+" = %d\n", props.LineLength)
	fmt.Printf(duration+" = %d\n", props.Duration)
	fmt.Printf(linesPerSecond+" = %d\n", props.Lps)
	fmt.Println(output + " =  %s" + props.FilePath)
	fmt.Printf(multiLinePercent+" = %d\n", props.MultiLinePercent)
	fmt.Printf(enableRotate+" = %t\n", props.Rotate)
	fmt.Printf(logRotationMaxSizeMegabytes+" = %d\n", props.RotateSizeMB)
}

/*
Accepts input commands and flags as arguments and the OS CLI
and binds to the relevant fields within the specified loadgenerator properties struct.
*/

func checkInsufficientArgs(cmd *cobra.Command, args []string) {
	if len(args) >= 1 && (args[0] == "-h" || args[0] == "--h" || args[0] == "--help") {
		cmd.Help()
		os.Exit(1)
	}
}

func Run(props *loadgen.LoadGenProperties) {

	var wg sync.WaitGroup
	wg.Add(1)
	props.Wg = &wg
	var metricsServerPortStr string
	var enableConfigVar bool
	promRegistry := prometheus.NewRegistry()

	rootCmd := &cobra.Command{
		Use:   "loadgen [COMMANDS]",
		Short: "loadgen is a tool which generates test-data (synthetic/replay) at a controlled rate to test logging infrastructure.",
		Run: func(cmd *cobra.Command, args []string) {
			checkInsufficientArgs(cmd, args)
			printInput(props)
		},
	}
	replayFileCmd := &cobra.Command{
		Use:   "replay-file",
		Short: "test data is generated by replaying a file.",
		Args: func(cmd *cobra.Command, args []string) error {
			flagSet := cmd.Flags()
			if flagSet.Changed(inputFilePath) == true {
				filePath := flagSet.Lookup(inputFilePath).Value.String()
				fmt.Println(filePath)
				info, err := os.Stat(filePath)
				if err != nil {
					return errors.New("File does not exist. input-file-path = " + filePath)
				}
				if info.Size() <= 0 {
					return errors.New("Invalid file size. filePath = " + filePath)
				}
			}

			//Check if multiple test termination criteria is defined.
			if flagSet.Changed(duration) && flagSet.Changed(replayCount) {
				return errors.New("Multiple test termination criteria found. You can either define --duration or --replay-count as termination criteria, not both. ")
			}
			return nil
		},
		Run: func(cmd *cobra.Command, args []string) {
			if props.EnableMetrics {
				go metrics.StartMetricsServer(promRegistry, &metricsServerPortStr)
			}
			setupFormatterWithTags(props, cmd)
			go loadgen.GenerateLoadFromInputFile(promRegistry, props)
			printInput(props)
			wg.Wait()
		},
	}
	randomStringsCmd := &cobra.Command{
		Use:   "random-strings",
		Short: "test data is generated by producing random alphanumeric strings.",
		Long:  "test data is generated by producing random alphanumeric strings.",
		Args: func(cmd *cobra.Command, args []string) error {
			flagSet := cmd.Flags()
			if ! flagSet.Changed(duration) {
				return errors.New("Duration is not specified. This command requires a termination criteria. Specify a duration (s) for this job. ")
			}
			return nil
		},
		Run: func(cmd *cobra.Command, args []string) {
			checkInsufficientArgs(cmd, args)
			if props.EnableMetrics {
				go metrics.StartMetricsServer(promRegistry, &metricsServerPortStr)
			}

			setupFormatterWithTags(props, cmd)
			go loadgen.GenerateAlphaNumeric(promRegistry, props)
			wg.Wait()
		},
	}

	//Defining persistent flags for root command
	rootCmd.PersistentFlags().Int64VarP(&props.Duration, duration, "d", constants.DefaultDurationInSeconds, "Duration of the job in seconds")
	rootCmd.PersistentFlags().Int64VarP(&props.Lps, linesPerSecond, "s", constants.DefaultLinesPerSecond, "target lines/second")
	rootCmd.PersistentFlags().Int64VarP(&props.FileCount, fileCount, "", constants.DefaultFileCount, "target requests/sec. Note: log rotation does not work in multi file scenario.")
	rootCmd.PersistentFlags().StringVarP(&cfgFile, config, "", "cfg", "-config /path/to/source/jsonfile")
	rootCmd.PersistentFlags().BoolVarP(&enableConfigVar, enableConfig, "", false, "use config file for configuration (default false)")
	rootCmd.PersistentFlags().StringVarP(&metricsServerPortStr, metricsServerPort, "p", constants.DefaultMetricsServerPort, "prometheus metrics server port. default is port 8080. ")
	rootCmd.PersistentFlags().BoolVarP(&props.EnableMetrics, enableMetrics, "", false, "true, enables prometheus metrics on http://<localhost>:<port>/metric HTTP endpoint for line emit rates and byte count. default is off.")
	rootCmd.PersistentFlags().StringVarP(&props.FilePath, output, "o", "", "file path of output log file ex: /var/tmp/test.log")
	rootCmd.PersistentFlags().BoolVarP(&props.Rotate, enableRotate, "r", false, "by default log rotation is off, use this flag to turn it on. Example --enable-log-rotation will turn log rotation on")
	rootCmd.PersistentFlags().Int64VarP(&props.RotateSizeMB, logRotationMaxSizeMegabytes, "", constants.DefaultMaxFileSize, "Rotate after N MegaBytes. example 100. ")
	rootCmd.PersistentFlags().StringVar(&props.Tags, tags, "t", "Custom hardcoded tags in the form of K=V")
	rootCmd.PersistentFlags().StringVar(&props.ResultLog, resultLog, "", "Path of the results log where metrics such as total_lines_generated will be log.")

	rootCmd.MarkPersistentFlagRequired(output)
	rootCmd.MarkPersistentFlagRequired(linesPerSecond)

	//Defining local flags for replay file command
	replayFileCmd.Flags().StringVarP(&props.InputSourceFile, inputFilePath, "i", "default.log", "file path of output log file ex: /var/tmp/test.log")
	replayFileCmd.MarkFlagRequired(inputFilePath)

	replayFileCmd.Flags().Int64VarP(&props.ReplayCount, replayCount, "", 0, "max number of replays. default is infinite, the replay stops when test duration expires.")

	//Defining local flags for random string generator command
	randomStringsCmd.Flags().IntVarP(&props.MultiLinePercent, multiLinePercent, "m", constants.DefaultMultiLinePercentage, " % of log entries which have multiple lines, example: 1% will generate ")
	randomStringsCmd.Flags().IntVarP(&props.NumOfLinesInMultiLineLog, lineCount, "n", constants.DefaultMultiLineCount, "max number of lines in multiline log entry")
	randomStringsCmd.Flags().Int64VarP(&props.LineLength, lineLength, "l", constants.DefaultLineLength, "line length(number of characters)")
	randomStringsCmd.MarkFlagRequired(lineCount)
	randomStringsCmd.MarkFlagRequired(lineLength)
	randomStringsCmd.MarkFlagRequired(multiLinePercent)

	rootCmd.AddCommand(randomStringsCmd, replayFileCmd)

	err := rootCmd.Execute()
	InputValidation(props)
	if err != nil {
		error.Error(err)
		os.Exit(1)
	}
}

func setupFormatterWithTags(props *loadgen.LoadGenProperties, cmd *cobra.Command) {
	flagSet := cmd.Flags()
	if flagSet.Changed(tags) {
		props.LogFormat = &easy.Formatter{
			LogFormat:       "%time% " + props.Tags + " %msg%\n",
			TimestampFormat: "2006-01-02T15:04:05.099-07:00",
		}
	}
}
